### ISSUE TITLE
Finish port to arm

### ISSUE THREAD

**Issue by mitchty** [mitchty/alpine-linux-ghc-bootstrap#8]
I had this almost working with 7.10.2 but ran into problems with the shared libraries not having been built right due to missing symbols in the haskell runtime shared library.

Figure out a way to get an arm port of 7.10 at least up. Been working on this off and on the past 6 months. Not in a huge rush. Help appreciated if anyone is ambitious, will push up a branch once I figure out what version kinda/sorta worked.

**Comment by mitchty** [comment]
Still working on this, in the past i managed to generate a stage 2 ghc 7.10.2 compiler. It didn't work but it built at least.

I just managed to finally build a stage 2 ghc 7.10.3, so we'll see if that actually holds up. I had to hack in and custom back port a lot of changes from ghc master to make things work. I also couldn't get integer-gmp to build in the stage2 to save my life. I'm pretty much over caring about getting integer-gmp to build into the bootstrap compiler.

Can watch progress (or normally lack thereof) here:
https://github.com/mitchty/alpine-linux-ghc-bootstrap/compare/llvm-3.6-experiment2...llvm-3.6-experiment2-arm?expand=1

Been working on this on/off for the past 10 months, so this is the most progress I've had to date. Have to go over the stage2 stuff with a fine comb to make sure it works but even just having configure && make succeed is a welcome step forward.

**Comment by mitchty** [comment]
Well I got this far overnight, this ghc stage 2 does at least build some things. But not itself yet.

```
[40 of 43] Compiling Data.Text.Lazy.Read ( Data/Text/Lazy/Read.hs, dist/build/Data/Text/Lazy/Read.o )

Data/Text/Lazy/Read.hs:3:14: Warning:
    ‘Data.Text.Lazy.Read’ is marked as Trustworthy but has been inferred as safe!

Data/Text/Lazy/Read.hs:32:1: Warning:
    The import of ‘Word’ from module ‘Data.Word’ is redundant
Call operand #8 has unhandled type i32
UNREACHABLE executed at /home/build/aports/testing/ghc/src/llvm-3.6.2.src/lib/CodeGen/CallingConvLower.cpp:131!
Stack dump:
0.      Program arguments: llc-3.6 -O3 -relocation-model=static /tmp/cabal-/ghc16289_0/ghc_459.bc -o /tmp/cabal-/ghc16289_0/ghc_460.lm_s --enable-tbaa=true -mattr=+v6 -float-abi=hard 
1.      Running pass 'Function Pass Manager' on module '/tmp/cabal-/ghc16289_0/ghc_459.bc'.
2.      Running pass 'ARM Instruction Selection' on function '@"rebX1_slow$def"'
Killed

Error during cabal-install bootstrap:
Building the text package failed.
>>> ERROR: ghc: all failed
>>> ghc: Uninstalling dependencies...
Command exited with non-zero status 1
real    3h 58m 14s
user    8h 19m 53s
sys     11m 9.35s
```

Not sure llvm-3.6 is a great choice here looking into that llvm #debug statement. First things I noticed now that I looked at gcc configuration on alpine arm is my cross compiler is too aggressive with architectures.

```
--with-arch=armv7-a --with-float=hard --with-fpu=vfpv3-d16
```

Should be

```
--with-arch=armv6 --with-float=hard --with-fpu=vfp
```

So I'm rebuilding the bootstrap compiler again, thats a 4 hour ordeal so I'm going to go for a run. Next up also need to figure out if llvm 3.5.2 will work or not with the recent patches I used to get the cross compiler working. A new experiment is in order, llvm 3.6 experiment might not survive arm porting.

I'll probably have to change the gcc-bootstrap build to use/build its own llvm as well assuming llvm 3.5.2 has to be used.

New things to try next/do:
- [x] Rebuild stage2 bootstrap compiler with the updated cross compiler gcc flags (in progress)
- [ ] Find a better way to get hp2hs and unlit built with the right C compiler (low priority, only unlit is really needed and easy to hack around, also I don't want to look at makefiles any more today)
- [x] Debug this llvm #debug a bit more after retrying with the updated bootstrap compiler
- [x] Experiment with llvm 3.5.2 again, hopefully thumb assembly doesn't show up in shared libraries as before
- [ ] Learn to meditate while waiting for compiles to finish https://xkcd.com/303/

**Comment by mitchty** [comment]
Well, so much for llvm 3.5, same error. The llvm folks seem to think its a bug. FUN! 

```
Data/Text/Lazy/Read.hs:3:14: Warning:
    ‘Data.Text.Lazy.Read’ is marked as Trustworthy but has been inferred as safe!

Data/Text/Lazy/Read.hs:32:1: Warning:
    The import of ‘Word’ from module ‘Data.Word’ is redundant
Call operand #8 has unhandled type i32
UNREACHABLE executed at /home/build/aports/testing/ghc/src/llvm-3.5.2.src/lib/CodeGen/CallingConvLower.cpp:129!
Stack dump:
0.      Program arguments: llc-3.5 -O3 -relocation-model=static /tmp/cabal-/ghc29607_0/ghc_459.bc -o /tmp/cabal-/ghc29607_0/ghc_460.lm_s --enable-tbaa=true -mattr=+v6 -float-abi=hard 
1.      Running pass 'Function Pass Manager' on module '/tmp/cabal-/ghc29607_0/ghc_459.bc'.
2.      Running pass 'ARM Instruction Selection' on function '@"rebX1_slow$def"'
Killed
+ die Building the text package failed.
+ printf \nError during cabal-install bootstrap:\nBuilding the text package failed.\n

Error during cabal-install bootstrap:
Building the text package failed.
+ exit 2
```

**Comment by mitchty** [comment]
Actually I wonder as it starts to snow/blizzard here if this llvm issue is somehow related to the config.sub hack for libffi to use musl-gcc here:
https://github.com/mitchty/alpine-linux-ghc-bootstrap/blob/1ed34759e28db99d7512d9a1693ddd907f0c8eb2/ghc-bootstrap/bootstrap/bootstrap-all.patch#L327

Always possible i'm holding it wrong and the ghc i've built is patently built incorrectly.

So two thoughts here for future testing, make that hack x86_64 only, OR just ditch the silly musl-gcc stuff entirely and build the x86_64 bootstrap compiler like I built the arm bootstrap compiler. Will test this later, letting 8.0 rc2 build for the moment.

**Comment by mitchty** [comment]
So looks like 8.0 will work with arm. Managed to build ghc 8.0 from git as of 3 days ago with the patches to fix llvm stuff.

```
# ghc --version
The Glorious Glasgow Haskell Compilation System, version 8.0.0.20160313
# ghci
GHCi, version 8.0.0.20160313: http://www.haskell.org/ghc/  :? for help
Prelude> :info Bool
data Bool = False | True        -- Defined in ‘GHC.Types’
instance Bounded Bool -- Defined in ‘GHC.Enum’
instance Enum Bool -- Defined in ‘GHC.Enum’
instance Eq Bool -- Defined in ‘GHC.Classes’
instance Ord Bool -- Defined in ‘GHC.Classes’
instance Read Bool -- Defined in ‘GHC.Read’
instance Show Bool -- Defined in ‘GHC.Show’
Prelude> 
```

**Comment by mitchty** [comment]
Forgot this bit too:

```
# file /usr/lib/ghc-8.0.0.20160313/bin/ghc
/usr/lib/ghc-8.0.0.20160313/bin/ghc: ELF 32-bit LSB shared object, ARM, EABI5 version 1 (SYSV), dynamically linked, interpreter /lib/ld-musl-armhf.so.1, stripped
```

Time to cleanup the patches and whatnot and check them in. Takes about 10 hours to build ghc, its not speedy.

**Comment by mitchty** [comment]
So as a note on this, finally pushed the work in progress up to this branch.
https://github.com/mitchty/alpine-linux-ghc-bootstrap/tree/ghc8.0-bootstrap

And if anyone wants to test the x86_64 stuff you can just docker run mitchty/alpine-ghc:8.0

Note I haven't gotten stack to build with all the changes yet so... no stack build off 8.0, I threw the 7.10.3 version into that repo cause I could.

The arm attempt that seems to work, modulo everything not expecting a new version of base etc.... is up at the same url too:
https://s3-us-west-2.amazonaws.com/alpine-ghc/next/8.0

Basically the same as x86_64 or the examples dir. It seems to work, haven't heavily tested it but it hasn't let me down so far. YMMV

But this will soon be what I push out for upstreaming once 8.0 is released. I'm giving up on 7.10.3 ghc on alpine arm mostly because I don't have the desire to figure out the llvm bugs.

**Comment by mitchty** [comment]
Right so mitchty/alpine-ghc:8.0 has ghc 8.0 for x86_64, as well as arm builds. They "seem to work" but no guarantee.

As an example, idris fails to build on 8.0 due to a weird parser issue relating to newlines that wasn't seen on 7.10.3 https://github.com/idris-lang/Idris-dev/issues/3127. So far the only issue I've seen in my testing though to be fair I don't have much to test with yet. Likely that for my next challenge I'll try to update the Idris parser to not depend on String or figure out what is broken there.

I submitted all this upstream (sans stack for the moment due to my abuse of --allow-newer to get it to build) to alpine linux as well so we will see how that goes or not. If I keep getting objections/pushback to my llvm solution I'll probably just work with the stack/ghc folks on getting a musl libc built ghc supported into stack or ghc themselves instead of getting this into alpine linux itself. I compiled stack and llvm fully statically so it wouldn't be that hard to get it supported on any musl libc distribution at this point for armhf or x86_64.

Closing this issue as most of what I can test works reasonably well for now, once more stuff actually gets updated to work with 8.0.1 I can find more bugs at that point. But for historical purposes people can peruse the old branches and see all the false starts at arm support with 7.10.3 and llvm etc... If anyone wants my notes on llvm 3.{5|6} and 7.10.3 and arm support to try getting that version up I can give it to them. But good luck there, lots of giant dragons abound in arm codegen paths.
